packages/viewer/src/lib/types/frame.ts

```diff
@@ -171,27 +201,38 @@ export class FramePerspectiveCamera
     return this.cameraMatrices;
   }
 
-  public get direction(): Vector3.Vector3 {
-    return Vector3.normalize(Vector3.subtract(this.lookAt, this.position));
-  }
-
-  public get worldMatrix(): Matrix4.Matrix4 {
-    return this.computeCameraMatrices().worldMatrix;
-  }
-
-  public get viewMatrix(): Matrix4.Matrix4 {
-    return this.computeCameraMatrices().viewMatrix;
-  }
-
-  public get projectionMatrix(): Matrix4.Matrix4 {
-    return this.computeCameraMatrices().projectionMatrix;
+  /**
+   * Checks if the given point, in world space, is behind the near plane
+   * of the camera.
+   *
+   * @param world A point in world space.
+   * @returns `true` if the point is behind the camera.
+   */
+  public isPointBehindNear(world: Vector3.Vector3): boolean {
+    const { position, direction, near } = this;
+    const dist = Vector3.dot(Vector3.subtract(world, position), direction);
+    return dist < near;
   }
 
-  public get projectionMatrixInverse(): Matrix4.Matrix4 {
-    return this.computeCameraMatrices().projectionMatrixInverse;
-  }
-
-  public get projectionViewMatrix(): Matrix4.Matrix4 {
-    return this.computeCameraMatrices().projectionViewMatrix;
+  /**
+   * Returns a point on the near plane that intersects with `line`. If `line`
+   * does not intersect, then `undefined` is returned.
+   *
+   * @param line The line to intersect.
+   * @returns A point in world space, or `undefined` if the line does not
+   * intersect with the near plane.
+   */
+  public intersectLineWithNear(line: Line3.Line3): Vector3.Vector3 | undefined {
+    const { position, lookAt, direction, near } = this;
+
+    const vs = Vector3.subtract(line.start, lookAt);
+    const ve = Vector3.subtract(line.end, lookAt);
+    const vl = Line3.create({ start: vs, end: ve });
+
+    const dist = Vector3.distance(lookAt, position) - near;
+    const nearP = Plane.create({ normal: direction, constant: dist });
```

Comment:
I'm thinking this should be `constant: -dist` since in the `Plane` functions, the `constant` is always added (hint: the plane is formed by all of the points where the sum of all the terms is zero so the `constant` needs to pull the sum in the opposite direction of everything else to make that happen).  This might explain why you were subtracting `lookAt` above (one sorta reverses the effect of the other), but I can also sometimes get myself confused.

Given that `dist` is a function of `position` (which intuitively makes total sense to me), it follows the rest of the vectors here should also be a function of `position` and not `lookAt`.

Reply:
My understanding of `constant` is that it's a distance from an origin. I guess origin could either be `position` or `lookAt`, but `lookAt` is a better mental model for me, as it's the center of rotation for the camera.

Reply:
Except `near` is defined relative to `position`, not `lookAt` so you're kinda mixing two coordinate systems together which happens to sorta work but I wouldn't depend on it.